# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1708979276 -3600
# Parent  db6bfb6e5fdb354f62a9ae224267dfcd853ed3fd
Feature: Bug 1831667: Re-establish counts, also show row count if different, take 2.

diff --git a/mail/base/content/about3Pane.js b/mail/base/content/about3Pane.js
--- a/mail/base/content/about3Pane.js
+++ b/mail/base/content/about3Pane.js
@@ -4816,17 +4816,18 @@ var threadPaneHeader = {
     document.l10n.setAttributes(
       this.folderCount,
       "thread-pane-folder-message-count",
       { count: gFolder?.getTotalMessages(false) || gDBView?.rowCount || 0 }
     );
 
     this.folderName.hidden = false;
     this.folderCount.hidden = false;
-    this.updateStatusBar();
+    // Spin the event loop to give view creation a breather.
+    setTimeout(() => this.updateStatusBar());
   },
 
   /**
    * Update the total message count in the header if the value changed for the
    * currently selected folder.
    *
    * @param {nsIMsgFolder} folder - The folder updating the count.
    * @param {integer} newValue
@@ -5112,16 +5113,17 @@ var threadPane = {
         break;
       case "expanded":
       case "collapsed":
         if (event.detail == threadTree.selectedIndex) {
           // The selected index hasn't changed, but a collapsed row represents
           // multiple messages, so for our purposes the selection has changed.
           threadTree.dispatchEvent(new CustomEvent("select"));
         }
+        threadPaneHeader.updateStatusBar();
         break;
       case "scroll":
         if (this.isFirstScroll) {
           this.isFirstScroll = false;
           break;
         }
         this.scrollDetected = true;
         break;
@@ -7524,16 +7526,17 @@ var sortController = {
         break;
       default:
         if (event.target.value in Ci.nsMsgViewSortType) {
           this.sortThreadPane(event.target.value);
           threadPane.restoreSortIndicator();
         }
         break;
     }
+    threadPaneHeader.updateStatusBar();
   },
   sortByThread() {
     threadPane.updateListRole(false);
     gViewWrapper.showThreaded = true;
     this.sortThreadPane("byDate");
   },
   sortThreadPane(sortName) {
     let sortType = Ci.nsMsgViewSortType[sortName];
@@ -7764,26 +7767,28 @@ commandController.registerCallback(
 
 commandController.registerCallback(
   "cmd_expandAllThreads",
   () => {
     threadPane.saveSelection();
     gViewWrapper.dbView.doCommand(Ci.nsMsgViewCommandType.expandAll);
     gViewWrapper._threadExpandAll = true;
     threadPane.restoreSelection();
+    threadPaneHeader.updateStatusBar();
   },
   () => !!gViewWrapper?.dbView
 );
 commandController.registerCallback(
   "cmd_collapseAllThreads",
   () => {
     threadPane.saveSelection();
     gViewWrapper.dbView.doCommand(Ci.nsMsgViewCommandType.collapseAll);
     gViewWrapper._threadExpandAll = false;
     threadPane.restoreSelection({ expand: false });
+    threadPaneHeader.updateStatusBar();
   },
   () => !!gViewWrapper?.dbView
 );
 
 function SwitchView(command) {
   // when switching thread views, we might be coming out of quick search
   // or a message view.
   // first set view picker to all
diff --git a/mail/base/content/mailCommon.js b/mail/base/content/mailCommon.js
--- a/mail/base/content/mailCommon.js
+++ b/mail/base/content/mailCommon.js
@@ -1052,16 +1052,17 @@ var dbViewWrapperListener = {
   onMessagesLoaded(all) {
     if (!window.threadPane) {
       return;
     }
 
     // There is no persisted thread last expanded state for synthetic views.
     if (all && !gViewWrapper.isSynthetic) {
       window.threadPane.restoreThreadState();
+      window.threadPaneHeader.updateStatusBar();
     }
 
     // Try to restore what was selected. Keep the saved selection (if there is
     // one) until we have all of the messages. This will also reveal selected
     // messages in collapsed threads.
     window.threadPane.restoreSelection({ discard: all });
 
     if (all || gViewWrapper.search.hasSearchTerms) {
